File: Lab 5/pi.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define CLOCKWISE   1
#define COUNTERCLK  2

int main(int argc, char *argv[]){
    int ierr;
    int rank, size;
    int src, dest;
    unsigned long iter; 
    unsigned long count = 0;
    int i;

    MPI_Status status;

    ierr = MPI_Init(&argc, &argv);
    ierr = MPI_Comm_size(MPI_COMM_WORLD, &size);
    ierr = MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    srand((unsigned int)time(NULL) * rank);

    if (rank == 0){
        iter = (long)atoi(argv[1]);
        
        for (int i = 1; i < size; i++){
            MPI_Send(&iter, 1, MPI_UNSIGNED_LONG, i, 0, MPI_COMM_WORLD);
        }
    } else {
        MPI_Recv(&iter, 1, MPI_UNSIGNED_LONG, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
    }

    for (i = 0; i < iter; i++){
        double x = (double)rand()  / ((double) RAND_MAX + 1.0);
        double y = (double)rand()  / ((double) RAND_MAX + 1.0);

        if (x*x + y*y < 1){
            count++;
        }
    }

    if (rank == 0){
        for (int i = 1; i < size; i++){
            unsigned long tmpCnt;
            MPI_Recv(&tmpCnt, 1, MPI_INT, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
            count += tmpCnt;
        }

        double pi = (double)4*((double)count/((double)size * (double)iter));

        printf("Approx of pi: %.9f\n", pi);

    } else {
         MPI_Send(&count, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
    }

    MPI_Finalize();
    return 0;
}

File: Lab 9/pi.f
      PROGRAM pi
      include 'mpif.h'
      INTEGER ierr
      INTEGER rank, size
      INTEGER src, dest
      INTEGER iter
      INTEGER :: count = 0
      REAL rand
      INTEGER i
      REAL x, y
      INTEGER tmpCount
      REAL PINum

      CALL MPI_INIT(ierr)
      CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierr)
      CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

      CALL RANDOM_SEED()

      IF (rank .eq. 0) THEN
            READ(*,*) iter

            DO i = 1, size-1
                  Call MPI_SEND(iter, 1, MPI_INTEGER, i, 0, MPI_COMM_WORLD, ierr)
            END DO      
      ELSE
            CALL MPI_RECV(iter, 1, MPI_INTEGER, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
      END IF

      DO i = 1, iter
            CALL RANDOM_NUMBER(rand)
            x = rand
            CALL RANDOM_NUMBER(rand)
            y = rand

            IF (x*x + y*y .LT. 1) THEN
                  count = count + 1
            END IF 
      END DO

      IF (rank == 0) THEN
            DO i=1,size-1
                  CALL MPI_Recv(tmpCount, 1, MPI_INTEGER, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
                  count = count + tmpCount
            END DO

            PINum = 4.0 * count / (size * iter)

            PRINT *, "PI: =", PINum

      ELSE 
            CALL MPI_Send(count, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, ierr)
      END IF

      CALL MPI_FINALIZE(ierr)
      
      END PROGRAM pi